### PR TITLE
PMM-11199: migrate pmm update to Grafana

### PR DESCRIPTION
**Pull request by fabio-silva**
**What this PR does / why we need it**: Moves a coded panel to Grafana, as planned. In this case, pmm-update.
Removed duplicated code in the way and changed tests from Enzyme to testing-library.

**Which issue(s) this PR fixes**: https://jira.percona.com/browse/PMM-11199

Related to https://github.com/percona/grafana-dashboards/pull/1460

ss

### CODE REVIEW

File: public/app/plugins/panel/pmm-update/UpdatePanel.tsx [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+          <div className={styles.middleSectionWrapper}>
+            <Spinner />
+          </div>
+        ) : (
+          <>
+            {(isUpdateAvailable || forceUpdate) && !updatesDisabled && !hasNoAccess && isOnline ? (
+              <div className={styles.middleSectionWrapper}>
+                {/* eslint-disable-next-line @typescript-eslint/consistent-type-assertions */}
+                <Button onClick={handleUpdate} icon={'fa fa-download' as IconName} variant="secondary">
+                  Upgrade to {nextVersionDetails?.nextVersion}
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/UpdatePanel.tsx:94]: Would it make sense to add it to the a messages file?

File: public/app/plugins/panel/pmm-update/UpdatePanel.utils.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { format } from 'date-fns';
+import { enUS } from 'date-fns/locale';
+
+import { ISOTimestamp } from './types';
+
+export const formatDateWithTime = (timestamp: ISOTimestamp) => {
+  const date = new Date(timestamp);
+  return format(new Date(date).valueOf() + date.getTimezoneOffset() * 60 * 1000, 'MMMM dd, H:mm', { locale: enUS });
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/UpdatePanel.utils.ts:8]: Is there a reason why a date is recreated? `date` is already a date object and it should be possible to call `valueOf` directly on it.

File: public/app/plugins/panel/pmm-update/UpdatePanel.utils.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+};
+
+export const formatDateWithYear = (timestamp: ISOTimestamp) => {
+  const date = new Date(timestamp);
+  return format(new Date(date).valueOf() + date.getTimezoneOffset() * 60 * 1000, 'MMMM dd, yyyy', { locale: enUS });
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/UpdatePanel.utils.ts:13]: Same as above.

File: public/app/plugins/panel/pmm-update/components/AvailableUpdate/AvailableUpdate.messages.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export const Messages = {
+  availableVersion: 'Available version',
+  tooltip: 'The date this version was published to the repository',
+  whatsNew: "What's new", // eslint-disable-line
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/components/AvailableUpdate/AvailableUpdate.messages.ts:4]: What's the `// eslint-disable-line` here for?

File: public/app/plugins/panel/pmm-update/components/CurrentVersion/CurrentVersion.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+export const CurrentVersion: FC<CurrentVersionProps> = ({ installedVersionDetails }) => {
+  const styles = useStyles(getStyles);
+  const [showFullVersion, handleToggleShowFullVersion] = useToggleOnAltClick(false);
+  const { installedVersionDate, installedVersion, installedFullVersion } = installedVersionDetails;
+
+  return (
+    <section className={styles.currentVersion}>
+      <span onClick={handleToggleShowFullVersion}>
+        {Messages.currentVersion}:{' '}
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/components/CurrentVersion/CurrentVersion.tsx:19]: Does the `{' '}` have the same function as the `nbsp` in the `AvailableUpdate` component?

File: public/app/plugins/panel/pmm-update/components/LastCheck/LastCheck.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import { Button, IconName } from '@grafana/ui';
+
+import { LastCheckProps } from '../../types';
+
+import * as styles from './LastCheck.styles';
+
+export const LastCheck: FC<LastCheckProps> = ({ lastCheckDate, onCheckForUpdates, disabled = false }) => (
+  <div className={styles.lastCheck}>
+    <p>
+      Last check: <span data-testid="update-last-check">{lastCheckDate}</span>
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/components/LastCheck/LastCheck.tsx:12]: Same as with the other component, would it make sense to have messages file here?

File: public/app/plugins/panel/pmm-update/components/ProgressModal/ProgressModal.styles.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+export const output = css`
+  height: 200px;
+  margin-right: 0;
+  margin-top: 15px;
+  overflow-y: scroll;
+  width: 100%;
+
+  pre {
+    background-color: #292929;
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/components/ProgressModal/ProgressModal.styles.ts:42]: Do the colors have to be hardcoded here? Doesn't it have issues in light mode?

File: public/app/plugins/panel/pmm-update/components/ProgressModal/ProgressModal.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        {!isUpdated ? (
+          <div className={styles.outputContent}>
+            <div className={styles.outputHeader}>
+              <Icon
+                className={styles.outputVisibilityToggle}
+                data-testid={`modal-chevron-icon-${chevronIcon}`}
+                name={chevronIcon}
+                onClick={handleToggleShowOutput}
+              />
+              <span>Log</span>
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/components/ProgressModal/ProgressModal.tsx:69]: Please add it to the messages.

File: public/app/plugins/panel/pmm-update/components/ProgressModal/ProgressModal.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+                {Messages.updateSuccessNotice} {version}
+              </h6>
+            </div>
+            <Button
+              className={styles.closeModal}
+              data-testid="modal-close"
+              variant="primary"
+              onClick={reloadAfterUpdate}
+            >
+              Close
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/components/ProgressModal/ProgressModal.tsx:100]: Same as with "Log".

File: public/app/plugins/panel/pmm-update/hooks/useClickOutside.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import { useEffect, RefObject } from 'react';
+
+export const useClickOutside = (ref: RefObject<HTMLElement>, handler: (e: Event) => void) => {
+  useEffect(() => {
+    const handleClick = (e: MouseEvent) => {
+      {
+        /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions */
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/hooks/useClickOutside.ts:7]: This comment is a bit weird here, a simple `//` comment should be fine.

File: public/app/plugins/panel/pmm-update/hooks/useInitializeUpdate.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        throw Error('Invalid response received');
+      }
+
+      const { auth_token, log_offset } = data;
+
+      setAuthToken(auth_token);
+      setInitialLogOffset(log_offset);
+    } catch (e) {
+      setUpdateFailed(true);
+      console.error(e);
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/hooks/useInitializeUpdate.ts:25]: We should probably log the error here with the logger.

File: public/app/plugins/panel/pmm-update/hooks/useVersionDetails.ts [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  });
+  const [lastCheckDate, setLastCheckDate] = useState('');
+  const [isUpdateAvailable, setIsUpdateAvailable] = useState(false);
+  const [data, errorMessage, isLoading, getVersionDetails] = useApiCall<GetUpdatesResponse | void, GetUpdatesBody>(
+    getCurrentVersion,
+    { force: initialForceUpdate },
+    { force: initialForceUpdate, onlyInstalledVersion: true }
+  );
+
+  // console.log(data, isLoading);
```
Comment (matejkubinec) [inline public/app/plugins/panel/pmm-update/hooks/useVersionDetails.ts:35]: Please remove.

**Review by matejkubinec — COMMENTED** [review]
Mainly just some small nitpicks.

**Comment by solovevayaroslavna** [comment]
I'm sorry, why we need it? I don't quite understand this from the ticket description
image

**Comment by fabio-silva** [comment]
> I'm sorry, why we need it? I don't quite understand this from the ticket description image

Our panels have lots of duplicated code (as in Grafana), logic, etc.
We had a discussion some time ago to move all coded panels to Grafana.
That way, it's easier to maintain these dashboards. Also, upgrading Grafana becomes easier, since it would be done mostly in one place. 
Same for CVEs on both repos.

File: public/app/plugins/panel/pmm-update/UpdatePanel.utils.ts (hunk 2 above)
Comment (fabio-silva) [reply public/app/plugins/panel/pmm-update/UpdatePanel.utils.ts]: Good one !

File: public/app/plugins/panel/pmm-update/components/CurrentVersion/CurrentVersion.tsx (hunk 5 above)
Comment (fabio-silva) [reply public/app/plugins/panel/pmm-update/components/CurrentVersion/CurrentVersion.tsx]: True. Changed to &nbsp

File: public/app/plugins/panel/pmm-update/components/ProgressModal/ProgressModal.styles.ts (hunk 7 above)
Comment (fabio-silva) [reply public/app/plugins/panel/pmm-update/components/ProgressModal/ProgressModal.styles.ts]: Well spotted. I got rid of hard-coded colors and used the theme

**Review by matejkubinec — DISMISSED** [review]

**Review by matejkubinec — DISMISSED** [review]

**Review by matejkubinec — APPROVED** [review]

**Pull request merged**